### PR TITLE
fuzz: Fix off by one error in sparse merkle trie item insertion

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Bug fix

**What does this PR do? Why is it needed?**

Off by one error can cause an out of bounds array panic.

Also refactor a bit of the validation to the constructor rather than putting validation in the struct methods.

**Which issues(s) does this PR fix?**

**Other notes for review**

Found with [fuzzbuzz.io](https://fuzzbuzz.io).

### CODE REVIEW

**Comment by nisdas** [comment]
Can we add a small unit test for this case ?

**Review by nisdas — CHANGES_REQUESTED** [review]
Also there is the same bug here, which isnt handled in the PR:
https://github.com/prysmaticlabs/prysm/pull/10662/files#diff-1e345f2ed9b4fe9fe77ad2b505742728465d2f4b894ca31704b05ef81f715141R163

**Comment by prestonvanloon** [comment]
> Can we add a small unit test for this case ?

The checked in fuzz test samples could count as a regression test. 
Do you think that is sufficient? They are run when you do `go test ./...`

**Comment by prestonvanloon** [comment]
> Also there is the same bug here, which isnt handled in the PR: https://github.com/prysmaticlabs/prysm/pull/10662/files#diff-1e345f2ed9b4fe9fe77ad2b505742728465d2f4b894ca31704b05ef81f715141R163

Yeah, I think all of these sanity checks can be done in the constructor. I'll refactor a bit for this PR

**Comment by prestonvanloon** [comment]
@nisdas I moved the validator into the constructor

File: container/trie/sparse_merkle.go [hunk 1]
```diff
@@ -39,7 +39,29 @@ func CreateTrieFromProto(trieObj *protodb.SparseMerkleTrie) *SparseMerkleTrie {
...
+
+func (m *SparseMerkleTrie) validate() error {
+	if len(m.branches) == 0 {
+		return errors.New("no branches")
+	}
+	if len(m.branches[len(m.branches)-1]) == 0 {
+		return errors.New("invalid branches provided")
+	}
+	if m.depth >= uint(len(m.branches)) {
+		return errors.New("depth is greater than number of branches")
```
Comment (nisdas) [inline container/trie/sparse_merkle.go:58]:
```suggestion
		return errors.New("depth is greater than or equal to number of branches")
```

File: container/trie/sparse_merkle_test.go [hunk 2]
```diff
@@ -16,6 +16,68 @@ import (
...
+				Depth: 2,
+			},
+			errString: "invalid branches provided",
+		},
+		{
+			trie: &ethpb.SparseMerkleTrie{
+				Layers: genValidLayers(3),
+				Depth:  12,
+			},
+			errString: "depth is greater than number of branches",
```
Comment (nisdas) [inline container/trie/sparse_merkle_test.go:63]:
```suggestion
			errString: "depth is greater than or equal to number of branches",
```

**Review by nisdas — APPROVED** [review]

**Pull request merged**